File: 07. Matrix/spiralMatrix.java
class Solution {
  
    public List<Integer> spiralOrder(int[][] mat) {
      //4 pointers to maintain the directions.
      
        int cl=0;
        int cr=mat[0].length-1;
        int rt=0;
        int rb=mat.length-1;
      
      //here i will be storing answers
      
        List<Integer> ans=new ArrayList<>();
      
      //x will be telling us where to top, bootom, left, right
        int x=0;
        while(cl<=cr && rt<=rb){
             x=x%4;
          //going left to right
            if(x==0){
                for(int i=cl;i<=cr;i++){
                    ans.add(mat[rt][i]);
                }
                rt++;
                x+=1;  
          //going right-top to bottom-right
              
            }else if(x==1){
                 for(int i=rt;i<=rb;i++){
                    ans.add(mat[i][cr]);
                }
                cr--;
                x+=1;
          //going bottom-right to bortom-left
                
            }else if(x==2){
               for(int i=cr;i>=cl;i--){
                    ans.add(mat[rb][i]);
                }
                rb--;
                x+=1;
                
          //going bottom-left to left-top
                
            }else if(x==3){
                 for(int i=rb;i>=rt;i--){
                    ans.add(mat[i][cl]);
                }
                cl++;
                x+=1;
                
                
            }
           
        }
        return ans;
        
    }
}

File: 07. Matrix/Spiral_Matrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        int m = matrix.size();
        int n = matrix[0].size();
        
        int top = 0, down = m-1;
        int left = 0, right = n-1;
        int dir = 0;
        vector<int> ans;
        
        
        while(top <= down and left <= right){
            if(dir==0){
                for(int i=left; i<=right; i++){
                    ans.push_back(matrix[top][i]);
                }
                top += 1;
            }
            else if(dir==1){
                for(int i=top; i<=down; i++){
                    ans.push_back(matrix[i][right]);
                }
                right -= 1;
            }
            else if(dir==2){
                for(int i=right; i>=left; i--){
                    ans.push_back(matrix[down][i]);
                }
                down -= 1;
            }
            else{
                for(int i=down; i>=top; i--){
                    ans.push_back(matrix[i][left]);
                }
                left += 1;
            }
            dir = (dir+1)%4;
        }
        
        return ans;
    }
};
